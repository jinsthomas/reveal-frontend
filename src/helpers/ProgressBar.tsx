import { Color } from 'csstype';
import React from 'react';
import * as colors from '../colors';
import { GREEN_THRESHOLD, ORANGE_THRESHOLD, YELLOW_THRESHOLD, ZERO } from '../configs/settings';

/** Marker? - a small vertical line on the progress bar that will visually inform of a certain limit */
interface Marker {
  height?: string /** customizable css style height */;
  width?: string /** how fat do you want it to be */;
  markAt: string /** percentage value as string; from left of progressbar that the marker should appear */;
  color?: Color /** shat color should the marker be */;
}

/** Props for ProgressBar */
interface ProgressBarProps {
  decimalPoints: number;
  height: string;
  min: number;
  max: number;
  value: number;
  marker: Marker | null;
}

/** default props for ProgressBar */
const defaultProgressBarProps: Partial<ProgressBarProps> = {
  decimalPoints: 0,
  height: '10px',
  max: 100,
  min: 0,
};

/** Displays configurable progress bar */
const ProgressBar = (props: ProgressBarProps) => {
  const { decimalPoints, height, value, marker } = props;
  const max = props.max || 100;
  const min = props.min || 0;
  let range = max - min;
  if (range <= 0) {
    range = 100;
  }
  const decimalValue = value / range;
  const percentValue = decimalValue * 100;
  const percentValueString = percentValue.toFixed(decimalPoints);

  return (
    <div className="progress" style={{ height, marginBottom: '15px', position: 'relative' }}>
      {!!marker && (
        <div
          className="marker"
          id="marker"
          style={{
            backgroundColor: marker.color ? marker.color : '#F00',
            height: marker.height ? marker.height : 'inherit',
            left: marker.markAt,
            position: 'absolute',
            width: marker.width ? marker.width : '5px',
          }}
        >
          &nbsp;
        </div>
      )}
      <div
        className={`progress-bar`}
        style={{
          backgroundColor:
            decimalValue >= GREEN_THRESHOLD
              ? colors.GREEN
              : decimalValue >= ORANGE_THRESHOLD
              ? colors.ORANGE
              : decimalValue >= YELLOW_THRESHOLD
              ? colors.RED
              : colors.YELLOW,
          width: `${percentValueString}%`,
        }}
        role="progressbar"
        aria-valuenow={percentValue}
        aria-valuemin={min}
        aria-valuemax={max}
      />
    </div>
  );
};

ProgressBar.defaultProps = defaultProgressBarProps;

export default ProgressBar;
