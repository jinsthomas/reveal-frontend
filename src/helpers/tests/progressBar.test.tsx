import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ProgressBar from '../ProgressBar';

describe('src/helpers/progressBar', () => {
  it('renders without crashing', () => {
    shallow(<ProgressBar />);
  });

  it('renders progressBar without marker as expected', () => {
    const wrapper = mount(<ProgressBar />);
    const markerWrapper = wrapper.find('#marker');
    expect(toJson(markerWrapper)).toMatchSnapshot('marker');
    expect(markerWrapper.length).toEqual(0);
  });

  it('renders markers correctly', () => {
    const props = {
      marker: {
        markAt: '20%',
      },
    };
    const wrapper = mount(<ProgressBar {...props} />);
    const markerWrapper = wrapper.find('#marker');
    expect(markerWrapper.length).toEqual(1);
    expect(toJson(markerWrapper)).toMatchSnapshot('marker');
    expect(markerWrapper.props().style).toEqual({
      backgroundColor: '#F00',
      height: 'inherit',
      left: '20%',
      position: 'absolute',
      width: '5px',
    });
  });
});
