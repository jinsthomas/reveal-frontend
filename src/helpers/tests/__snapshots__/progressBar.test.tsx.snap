// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/helpers/progressBar renders markers correctly: marker 1`] = `
<div
  className="marker"
  id="marker"
  style={
    Object {
      "backgroundColor": "#F00",
      "height": "25px",
      "left": "20%",
      "position": "absolute",
      "top": "-5px",
      "width": "5px",
      "zIndex": 1,
    }
  }
>
  Â 
</div>
`;

exports[`src/helpers/progressBar renders progressBar without marker as expected: marker 1`] = `null`;
